compiler/rustc_data_structures/src/vec_cache.rs

```diff
@@ -0,0 +1,266 @@
+//! VecCache maintains a mapping from K -> (V, I) pairing. K and I must be roughly u32-sized, and V
+//! must be Copy.
+//!
+//! VecCache supports efficient concurrent put/get across the key space, with write-once semantics
+//! (i.e., a given key can only be put once). Subsequent puts will panic.
+//!
+//! This is currently used for query caching.
+
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::mem::offset_of;
+use std::sync::atomic::{AtomicPtr, AtomicU32, Ordering};
+
+use rustc_index::Idx;
+
+use crate::sync::AtomicUsize;
+
+struct Slot<V> {
+    // We never construct &Slot<V> so it's fine for this to not be in an UnsafeCell.
+    value: V,
+    // This is both an index and a once-lock.
+    //
+    // 0: not yet initialized.
+    // 1: lock held, initializing.
+    // 2..u32::MAX - 2: initialized.
+    index_and_lock: AtomicU32,
+}
+
+impl<K: Idx, V, I> Default for VecCache<K, V, I> {
+    fn default() -> Self {
+        VecCache {
+            buckets: Default::default(),
+            key: PhantomData,
+            len: Default::default(),
+            present: Default::default(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+struct SlotIndex {
+    bucket_idx: usize,
+    entries: usize,
+    index_in_bucket: usize,
+}
```

Comment:
please add a comment to either the struct definition or to `from_index` which explain how this works, even  if it duplicates the comment on `VecCache` itself